src/server/scripts/Spells/spell_druid.cpp

```diff
@@ -1386,6 +1389,99 @@ class spell_dru_thrash_aura : public AuraScript
     }
 };
 
+// 740 - Tranquility
+class spell_dru_tranquility : public SpellScript
+{
+    void HandleHitTarget(SpellEffIndex effIndex)
+    {
+        Unit* caster = GetCaster();
+
+        // Note: Inner Peace talent.
+        if (!caster->HasAura(SPELL_DRUID_INNER_PEACE))
+            PreventHitDefaultEffect(effIndex);
+    }
+
+    void Register() override
+    {
+        OnEffectHitTarget += SpellEffectFn(spell_dru_tranquility::HandleHitTarget, EFFECT_3, SPELL_EFFECT_APPLY_AURA);
+        OnEffectHitTarget += SpellEffectFn(spell_dru_tranquility::HandleHitTarget, EFFECT_4, SPELL_EFFECT_APPLY_AURA);
+    }
+};
+
+// 157982 - Tranquility (Heal)
+class spell_dru_tranquility_heal : public SpellScript
+{
+    bool Validate(SpellInfo const* /*spellInfo*/) override
+    {
+        return ValidateSpellEffect({ { SPELL_DRUID_TRANQUILITY, EFFECT_2 } });
+    }
+
+    void FilterTargets(std::list<WorldObject*>& targets)
+    {
+        uint32 const maxTargets = uint32(GetSpellInfo()->MaxAffectedTargets);
+
+        // Note: Tranquility became a smart heal which prioritizes players and their pets in their group before any unit outside their group.
+        Trinity::SelectRandomInjuredTargets(targets, maxTargets, true);
+    }
+
+    void HandleCalculateHeal(SpellEffIndex /*effIndex*/)
+    {
+        Player* player = GetCaster()->ToPlayer();
+        if (!player)
+            return;
+
+        float healBonus = 1.0f;
+
+        // Note: if caster doesn't have group or has one which is not a raid group.
+        Group const* group = player->GetGroup();
+        if (!group || (group && !group->isRaidGroup()))
+        {
+            if (AuraEffect* tranquilityEffect = player->GetAuraEffect(SPELL_DRUID_TRANQUILITY, EFFECT_2))
+                AddPct(healBonus, tranquilityEffect->GetAmount());
+        }
+
+        SetHitHeal(GetHitHeal() * healBonus);
+    }
+
+    void Register() override
+    {
+        OnObjectAreaTargetSelect += SpellObjectAreaTargetSelectFn(spell_dru_tranquility_heal::FilterTargets, EFFECT_ALL, TARGET_UNIT_SRC_AREA_ALLY);
+        OnEffectHitTarget += SpellEffectFn(spell_dru_tranquility_heal::HandleCalculateHeal, EFFECT_0, SPELL_EFFECT_HEAL);
+    }
+};
+
+class spell_dru_tranquility_heal_aura : public AuraScript
+{
+    bool Validate(SpellInfo const* /*spellInfo*/) override
+    {
+        return ValidateSpellEffect({ { SPELL_DRUID_TRANQUILITY, EFFECT_2 } });
+    }
+
+    void HandleCalculateAmount(AuraEffect const* /*aurEff*/, int32& amount, bool& /*canBeRecalculated*/)
+    {
+        Player* player = GetCaster()->ToPlayer();
+        if (!player)
+            return;
+
+        float healBonus = 1.0f;
+
+        // Note: if caster doesn't have group or has one which is not a raid group.
+        Group const* group = player->GetGroup();
+        if (!group || (group && !group->isRaidGroup()))
```

Comment:
You can create a base class for just the bonus calculation so you dont have them duplicated.

As example:

```
class Test
{
    void ApplyTranqBonus(Player* player, float& healBonus)
    {
        Group const* group = player->GetGroup();
        if (!group || (group && !group->isRaidGroup()))
        {
            if (AuraEffect* tranquilityEffect = player->GetAuraEffect(SPELL_DRUID_TRANQUILITY, EFFECT_2))
                AddPct(healBonus, tranquilityEffect->GetAmount());
        }
    }
};

class spell_dru_tranquility_heal_aura : public AuraScript, Test
{
}
```